### DISCUSSION TITLE
Is it possible to use real parameters to fit dispersive material?

### DISCUSSION THREAD

**Discussion by Social-Mean** [flexcompute/tidy3d#1528] [Q&A] [answered]
Hi everyone, I have read the source code in the fit material part. This project use CCPRP to fit dispersive material, and the fit parameters are complex number. When using these parameters in FDTD, it will be slow (because one complex number multiplication is equivalent to four real number multiplications).

So I am asking is it possible to use real parameters to fit dispersive material? Or, how to use complex numbers for FDTD calculations without loss of efficiency?

Any help will be appreciated, thank you!

**Comment by tylerflex** [comment]
@weiliangjin2021 ?

**Comment by momchil-flex** [comment]
If there is, we would certainly like to know about it. :)

FDTD is actually memory bound. So thinking about it in terms of "one complex number multiplication is equivalent to four real number multiplications" ins't really correct. The question is how many auxiliary currents you need to read and write from memory, per grid point in a dispersive medium, per pole. The innovation of the CCPR paper was that it is a more general approach than e.g. Lorentzian poles, while requiring the same number of auxiliary currents. So it is more expressive at the same cost. But yeah, we are not aware of anything better.

**Reply by Social-Mean** [reply]
@momchil-flex Thanks for your reply.

> The question is how many auxiliary currents you need to read and write from memory, per grid point in a dispersive medium, per pole.

The formula to calculate auxiliary currents is:
$$\mathbf{J}_p^{(n+1)\Delta t} = k_p\mathbf{J}_p^{n\Delta t} + \beta_p\left(\frac{\mathbf{E}^{(n+1)\Delta t} - \mathbf{E}^{n\Delta t}}{\Delta t}\right),$$
where $k_p$ and $\beta_p$ are complex number. The count of auxiliary currents is related to the density of the grid. When densities of grid are same, CCPRP model take more time than multipole Lorentz model because of complex number.

I think we need more method in material fitting, like fitting multipole Debye model, or multipole Lorentz model, which is contained by other commercial softwares like CST, COMSOL and Ansys Lumerical FDTD. But I have not found any papers focusing on how to fit these model correctly like vector fitting used in CCPRP model.

**Reply by momchil-flex** [reply, marked as answer]
I understand what you are driving at, but really it turns out that FLOPs basically don't matter. All that matters is how many reads and writes from memory you need to do. This is always true on GPU and also on CPU if using a parallel FDTD that utilizes all cores. The only case in which you will be compute limited instead of memory limited (in which case the number of multiplications / additions you need to do will matter) is if you're running e.g. on a single CPU core and utilizing the entire CPU bandwidth. We actually have a draft of an article explaining some of this, but it's not really new knowledge, it has been discussed in other sources too.

So the answer to your question is already given in the Han and Fan paper that introduced the CCPR treatment:

![image](https://github.com/flexcompute/tidy3d/assets/92756559/851dd316-ce15-4ed1-a099-64b7a1252da2)

Basically, for the same number of poles P, a Lorentz or Debye fit will be equally fast (or slow), but the fit is less expressive, so it will likely be worse. Now, *maybe* there are some situations in which a user might still prefer it, although it's hard for me to come up with such a case.

**Reply by Social-Mean** [reply]
@momchil-flex I understand. Thank you for your detailed and attentive replies. 😃
